### PR TITLE
flow controller: consider the time factor when evaluating the trending

### LINKED ISSUE #11530: flow_controller.rs trend function didn't consider the time range
## Development Task

- [x] The flow_controller.rs trend function didn't consider the time range, so it maybe not accurate when evaluating the trending of pending compaction bytes or pending memtable numbers.

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Signed-off-by: zhangjinpeng1987 <zhangjinpeng@pingcap.com>

### What is changed and how it works?

Close #11530

What's Changed:

Consider time factor when evaluating the trending of pending compaction bytes or pending compaction memtable numbers.

### Related changes

- No

### Check List 

Tests 

- Unit test

Side effects

- No

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/storage/txn/flow_controller.rs [hunk 1]
```diff
@@ -345,20 +375,22 @@ struct CFFlowChecker {
...
     last_flush_bytes: u64,
     last_flush_bytes_time: Instant,
-    short_term_l0_production_flow: Smoother<u64, 10>,
+    short_term_l0_production_flow: Smoother<u64, 512>,
 
     // L0 consumption flow related
     last_l0_bytes: u64,
     last_l0_bytes_time: Instant,
-    short_term_l0_consumption_flow: Smoother<u64, 3>,
+    short_term_l0_consumption_flow: Smoother<u64, 512>,
```
Comment (Connor1996) [inline src/storage/txn/flow_controller.rs:388]:
```suggestion
    long_term_l0_consumption_flow: Smoother<u64, 512>,
```

File: src/storage/txn/flow_controller.rs [hunk 2]
```diff
@@ -345,20 +375,22 @@ struct CFFlowChecker {
...
     // So to evaluate the accumulation of L0 files, here only records the number
     // of L0 files right after L0 compactions.
-    long_term_num_l0_files: Smoother<u64, 20>,
+    long_term_num_l0_files: Smoother<u64, 512>,
 
     // L0 production flow related
     last_flush_bytes: u64,
     last_flush_bytes_time: Instant,
-    short_term_l0_production_flow: Smoother<u64, 10>,
+    short_term_l0_production_flow: Smoother<u64, 512>,
```
Comment (Connor1996) [inline src/storage/txn/flow_controller.rs:383]:
```suggestion
    long_term_l0_production_flow: Smoother<u64, 512>,
```

File: src/storage/txn/flow_controller.rs [hunk 3]
```diff
@@ -334,7 +359,12 @@ where
 // And all of them are collected from the hook of RocksDB's event listener.
 struct CFFlowChecker {
     // Memtable related
-    last_num_memtables: Smoother<u64, 20>,
+    // Assume the write flow is about 100MB/s, since the memtable size of
+    // write & default cf is 128MB, so there are about 300 flush events in 5
+    // minutes, we will observe the number of memtables about 300 times. We
+    // set the capacity of `last_num_memtables` to 512 to make sure we keep
+    // enough fresh record in the most cases.
+    last_num_memtables: Smoother<u64, 512>,
```
Comment (Connor1996) [inline src/storage/txn/flow_controller.rs:367]: large capacity would make the average of num_memtables lagging, which can't avoid the accumulation of memtables in time.

File: src/storage/txn/flow_controller.rs [hunk 4]
```diff
@@ -345,20 +375,22 @@ struct CFFlowChecker {
...
     last_l0_bytes: u64,
     last_l0_bytes_time: Instant,
-    short_term_l0_consumption_flow: Smoother<u64, 3>,
+    short_term_l0_consumption_flow: Smoother<u64, 512>,
 
     // Pending compaction bytes related
-    long_term_pending_bytes: Smoother<f64, 60>,
+    // When the write flow is about 100MB/s, we observed that the compaction ops
+    // is about 2.5, it means there are 750 compaction events in 5 minutes.
+    long_term_pending_bytes: Smoother<f64, 1024>,
```
Comment (Connor1996) [inline src/storage/txn/flow_controller.rs:393]: ditto

File: src/storage/txn/flow_controller.rs [hunk 5]
```diff
@@ -236,23 +239,25 @@ where
         + FromPrimitive,
 {
     pub fn observe(&mut self, record: T) {
+        self.observe_with_time(record, Instant::now_coarse());
+    }
+
+    pub fn observe_with_time(&mut self, record: T, time: Instant) {
```
Comment (tabokie) [inline src/storage/txn/flow_controller.rs:245]:
```suggestion
    fn observe_with_time(&mut self, record: T, time: Instant) {
```

File: src/storage/txn/flow_controller.rs [hunk 6]
```diff
@@ -224,7 +226,8 @@ where
     }
 }
 
-impl<T, const CAP: usize> Smoother<T, CAP>
+impl<T, const CAP: usize, const STALE_DUR: u64, const MIN_TIME_SPAN: u64>
```
Comment (tabokie) [inline src/storage/txn/flow_controller.rs:229]: How about
```suggestion
impl<T, const CAP: usize, const MAX_SPAN_SECS: u64, const MIN_SPAN_SECS: u64>
```

**Review by tabokie — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**